dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -1069,6 +1116,41 @@ private static ChatRequestMessage GetRequestMessage(ChatResponseMessage message)
         throw new NotSupportedException($"Role {message.Role} is not supported.");
     }
 
+    private OpenAIChatMessageContent GetChatMessage(ChatChoice chatChoice, ChatCompletions responseData)
+    {
+        var message = new OpenAIChatMessageContent(chatChoice.Message, this.DeploymentOrModelName, GetChatChoiceMetadata(responseData, chatChoice));
+
+        foreach (var toolCall in chatChoice.Message.ToolCalls)
+        {
+            // Adding items of 'FunctionCallContent' type to the 'Items' collection even though the function calls are available via the 'ToolCalls' property.
+            // This allows consumers to work with functions in an LLM-agnostic way.
+            if (toolCall is ChatCompletionsFunctionToolCall functionToolCall)
+            {
+                KernelArguments? arguments = null;
+                try
+                {
+                    arguments = JsonSerializer.Deserialize<KernelArguments>(functionToolCall.Arguments);
+                }
+                catch (JsonException)
+                {
+                    // If the arguments are not valid JSON, we'll just leave them as null.
+                    // The original arguments and function tool call will be available via the 'InnerContent' property for the connector caller to access.
```

Comment:
We need to return the exception to the caller, whether it's an LLM or a connector caller. The code is not there yet; I'm considering its design.  
   
This method iterates over LLM tool calls and converts each to a `FunctionCallRequestContent` item that contains all information about the function - plugin/function name, id, and arguments. As part of the conversion, function arguments are deserialized to kernel arguments. If the arguments' deserialization fails, the exception needs to be saved (to be replayed later when a caller invokes the function) in the `FunctionCallRequestContent` item, and the conversion should move to the next tool rather than stopping the whole operation just because one has invalid arguments.

An argument can be made in favor of stopping the entire operation if LLM provided invalid arguments for one or a few tools. As a counterargument, the caller may have code in place to fix those arguments, so the whole operation should not be stopped.  
   
If it's decided to proceed with the operation despite invalid arguments for a few functions, then it's up to the caller what to do with the exception. It can be converted to "Error: Failed to deserialize function arguments" and sent to LLM, or it can be re-thrown as any other .Net exception if the caller was not aware of it and does not have code for handling it.  
   
At the moment, there are a few ways the exception can be saved in the `FunctionCallRequestContent` item:  
1. In the `FunctionCallRequestContent.Exception` property, which is not there and needs to be added.  
2. In the `FunctionCallRequestContent.Metadata` with a well-known key.  
3. Passed to the `FunctionCallRequestContent` item constructor instead of `KernelArguments` and thrown when the function arguments are accessed by a caller.  
   
Additionally, the serialization and deserialization aspect of chat history, which the `FunctionCallRequestContent` item may be part of, should be taken into account. It's not easy/possible to deserialize a serialized exception back to the original type without knowing original exception type so an alternative solution needs to be found. InvalidKernelFunctionArgumentsException?

Reply:
Decided to go with option 1 because it's a simple solution for the edge-case scenario.